pkg/s3wrapper/upload_iso.go

```diff
@@ -0,0 +1,311 @@
+	if err != nil {
+		return err
+	}
+	completedParts = append(completedParts, completedPart)
+
+	log.Debugf("Completing multi-part upload of ISO %s", objectName)
+	_, err = u.s3client.CompleteMultipartUpload(&s3.CompleteMultipartUploadInput{
+		Bucket:   aws.String(u.bucket),
+		Key:      aws.String(objectName),
+		UploadId: uploadID,
+		MultipartUpload: &s3.CompletedMultipartUpload{
+			Parts: completedParts,
+		},
+	})
+	if err != nil {
+		err = errors.Wrapf(err, "Failed to complete upload for %s", objectName)
+		log.Error(err)
+		return err
+	}
+	log.Debugf("Completed upload of ISO %s", objectName)
+	return nil
+}
+
+func (u *ISOUploader) getISOInfo(baseObjectName string, log logrus.FieldLogger) (*isoInfo, *[]byte, error) {
+	var info isoInfo
+	var origContents []byte
+
+	// Get ETag from S3
+	headResp, err := u.s3client.HeadObject(&s3.HeadObjectInput{
+		Bucket: aws.String(u.bucket),
+		Key:    aws.String(baseObjectName),
+	})
+	if err != nil {
+		log.WithError(err).Errorf("Failed to fetch metadata for base object %s", baseObjectName)
+		return nil, nil, err
+	}
+
+	// See if the ISO info is cached
+	var found *isoInfo
+	for _, i := range u.infoCache {
+		if i.etag == *headResp.ETag {
+			found = &i
+		}
+	}
+	cachePath := filepath.Join("/tmp", *headResp.ETag)
+
+	if found == nil {
+		// Add to cache if not found
+		log.Debugf("Did not find ISO info for %s in cache", baseObjectName)
+		var offset, length int64
+		offset, length, err = u.getISOHeaderInfo(log, baseObjectName, *headResp.ContentLength)
+		if err != nil {
+			err = errors.Wrapf(err, "Failed to get base ISO info for %s from S3", baseObjectName)
+			log.Error(err)
+			return nil, nil, err
+		}
+		info.etag = *headResp.ETag
+		info.baseObjectSize = *headResp.ContentLength
+		info.areaLengthBytes = length
+		info.areaOffsetBytes = offset
+
+		var getRest *s3.GetObjectOutput
+		getRest, err = u.s3client.GetObject(&s3.GetObjectInput{
+			Bucket: aws.String(u.bucket),
+			Key:    aws.String(baseObjectName),
+			Range:  aws.String(fmt.Sprintf("bytes=%d-%d", offset, offset+minimumPartSizeBytes-1)),
+		})
+		if err != nil {
+			err = errors.Wrapf(err, "Failed to fetch embedded area of live ISO %s", baseObjectName)
+			log.Error(err)
+			return nil, nil, err
+		}
+		origContents, err = ioutil.ReadAll(getRest.Body)
+		if err != nil {
+			err = errors.Wrapf(err, "Failed to fetch body from embedded area of live ISO %s", baseObjectName)
+			log.Error(err)
+			return nil, nil, err
+		}
+
+		err = ioutil.WriteFile(cachePath, origContents, 0644)
+		if err != nil {
+			log.WithError(err).Errorf("Failed to cache embedded area to file %s", cachePath)
+		} else {
+			u.infoCache = append(u.infoCache, info)
+		}
+	} else {
+		// Return from cache
+		log.Debugf("Found ISO info for %s in cache", baseObjectName)
+		info = *found
+		origContents, err = ioutil.ReadFile(cachePath)
+		if err != nil {
```

Comment:
What will happen if we get the info but return empty string - we will try to upload part with empty content? 
Maybe we should return an error?